tools/benchmark/cmd/watch_latency.go

```diff
@@ -57,54 +63,90 @@ func init() {
 func watchLatencyFunc(_ *cobra.Command, _ []string) {
 	key := string(mustRandBytes(watchLKeySize))
 	value := string(mustRandBytes(watchLValueSize))
-
-	clients := mustCreateClients(totalClients, totalConns)
+	wchs := setupWatchChannels(key)
 	putClient := mustCreateConn()
 
-	wchs := make([]clientv3.WatchChan, len(clients))
-	for i := range wchs {
-		wchs[i] = clients[i].Watch(context.TODO(), key)
-	}
-
-	bar = pb.New(watchLTotal)
+	bar = pb.New(watchLPutTotal * len(wchs))
 	bar.Start()
 
 	limiter := rate.NewLimiter(rate.Limit(watchLPutRate), watchLPutRate)
-	r := newReport()
-	rc := r.Run()
 
-	for i := 0; i < watchLTotal; i++ {
+	putTimes := make([]time.Time, watchLPutTotal)
+	eventTimes := make([][]time.Time, len(wchs))
+
+	for i, wch := range wchs {
+		wch := wch
+		i := i
+		eventTimes[i] = make([]time.Time, watchLPutTotal)
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			eventCount := 0
+			for eventCount < watchLPutTotal {
+				resp := <-wch
+				for range resp.Events {
+					eventTimes[i][eventCount] = time.Now()
+					eventCount++
+					bar.Increment()
+				}
+			}
+		}()
+	}
+
+	putReport := newReport()
+	putReportResults := putReport.Run()
+	watchReport := newReport()
+	watchReportResults := watchReport.Run()
+	for i := 0; i < watchLPutTotal; i++ {
 		// limit key put as per reqRate
 		if err := limiter.Wait(context.TODO()); err != nil {
 			break
 		}
-
-		var st time.Time
-		var wg sync.WaitGroup
-		wg.Add(len(clients))
-		barrierc := make(chan struct{})
-		for _, wch := range wchs {
-			ch := wch
-			go func() {
-				<-barrierc
-				<-ch
-				r.Results() <- report.Result{Start: st, End: time.Now()}
-				wg.Done()
-			}()
-		}
-
+		start := time.Now()
 		if _, err := putClient.Put(context.TODO(), key, value); err != nil {
 			fmt.Fprintf(os.Stderr, "Failed to Put for watch latency benchmark: %v\n", err)
 			os.Exit(1)
 		}
-
-		st = time.Now()
-		close(barrierc)
-		wg.Wait()
-		bar.Increment()
+		end := time.Now()
+		putReport.Results() <- report.Result{Start: start, End: end}
+		putTimes[i] = end
 	}
-
-	close(r.Results())
+	wg.Wait()
+	close(putReport.Results())
 	bar.Finish()
-	fmt.Printf("%s", <-rc)
+	fmt.Printf("\nPut summary:\n%s", <-putReportResults)
+
+	for i := 0; i < len(wchs); i++ {
+		for j := 0; j < watchLPutTotal; j++ {
+			start := putTimes[j]
+			end := eventTimes[i][j]
+			if end.Before(start) {
+				start = end
+			}
+			watchReport.Results() <- report.Result{Start: start, End: end}
+		}
+	}
+
+	close(watchReport.Results())
+	fmt.Printf("\nWatch events summary:\n%s", <-watchReportResults)
+}
+
+func setupWatchChannels(key string) []clientv3.WatchChan {
+	clients := mustCreateClients(totalClients, totalConns)
+
+	streams := make([]clientv3.Watcher, watchLStreams)
```

Comment:
It seems that it doesn't really create `watchLStreams` gRPC streams. Instead it just create `totalConns` gRPC streams?

Reply:
by streams here I mean `watch streams` aka `Watch RPC`. I based the code based on tools/benchmark/cmd/watch.go where I discovered that you can separate watch requests into different stream by creating a new watcher. This reuses same grpc connection as you pass it the same etcd client,  however creates a separate stream as those are stored on watcher. Seems like a better way to separate streams approach I proposed in Kubernetes via grpc context metadata.